### PR TITLE
Training Loss Evaluation

### PR DESCRIPTION
**Pull request by rufex2001**
This PR adds an evaluation job that is simply the training loss. Currently, support is only for computing the loss on the training split. This is part of the following PR:

https://github.com/uma-pi1/kge/pull/99

### CODE REVIEW

File: kge/job/train.py [hunk 1]
```diff
@@ -78,15 +81,22 @@ def __init__(
...
-        if self.config.get("valid.split") != "":
-            valid_conf.set("eval.split", self.config.get("valid.split"))
-        valid_conf.set("eval.trace_level", self.config.get("valid.trace_level"))
-        self.valid_job = EvaluationJob.create(
-            valid_conf, dataset, parent_job=self, model=self.model
-        )
+        self.is_forward_only = forward_only
+
+        if not self.is_forward_only:
+            self.model.train()
```
Comment (rgemulla) [inline kge/job/train.py:87]: otherwise need to set self.model.eval()? Or is this done outside in the surrounding evaluation job?

File: kge/job/train.py [hunk 2]
```diff
@@ -316,8 +344,12 @@ def run_epoch(self) -> Dict[str, Any]:
                 "split": self.train_split,
                 "batch": batch_index,
                 "batches": len(self.loader),
-                "lr": [group["lr"] for group in self.optimizer.param_groups],
+                # "lr": [group["lr"] for group in self.optimizer.param_groups],
```
Comment (rgemulla) [inline kge/job/train.py:347]: Don't comment, remove.

File: kge/job/train.py [hunk 3]
```diff
@@ -856,13 +891,30 @@ def _process_subbatch(
...
+        # But run() is never called on a training job
+        # used for a training loss eval job
+        # Run() is called on the eval job, which calls
+        # run_epoch on its training job
+        # TODO not nice but prepare methods are private
+        #   so a nicer solution requires more change
+        if not self._is_prepared:
+            super()._prepare()
+            self._prepare()
+            self._is_prepared = True
```
Comment (rgemulla) [inline kge/job/train.py:910]: This should not be done here, but outside in TrainingEvalJob

File: kge/job/train.py [hunk 4]
```diff
@@ -1002,15 +1054,32 @@ def _process_subbatch(
...
+
+        return super().run_epoch()
```
Comment (rgemulla) [inline kge/job/train.py:1075]: This would also avoid these unnecessary repetitions.

**Comment by rgemulla** [comment]
Also, the configuraiton should be updated (saying that there is now this other evaluation type).

File: kge/job/train.py (hunk 1 above)
Comment (rufex2001) [reply kge/job/train.py:87]: Done in the eval job

**Comment by rufex2001** [comment]
@rgemulla All comments addressed in latest commit

**Comment by rufex2001** [comment]
@rgemulla Please wait until I add sum_loss to tracing for merging

**Comment by rufex2001** [comment]
@rgemulla I've added sum loss tracing to trainers and the training loss eval job. Two things:

1. Should we do this as simple as it is how? Or should we add this also at the batch trace level? This would mean trainers add sum_loss to their results object.
2. The solution to preparing the train job when creating it as an eval job (eval.py line 172) calls a private method from outside. Not sure this is what you meant, but this is partly why I didn't do it like this directly, but I forgot to bring it up (until now).

File: kge/job/eval.py [hunk 5]
```diff
@@ -141,3 +146,66 @@ def create_from(
             new_config.set("eval.split", eval_split, create=True)
 
         return super().create_from(checkpoint, new_config, dataset, parent_job)
+
+
+class TrainingLossEvaluationJob(EvaluationJob):
+    """ Evaluating simply by using the training loss """
```
Comment (rgemulla) [inline kge/job/eval.py:152]: Drop "simply" ;)

File: kge/job/eval.py [hunk 6]
```diff
@@ -141,3 +146,66 @@ def create_from(
...
+
+    def __init__(self, config: Config, dataset: Dataset, parent_job, model):
+        super().__init__(config, dataset, parent_job, model)
+
+        training_loss_eval_config = config.clone()
+        training_loss_eval_config.set("console.quiet", False)
```
Comment (rgemulla) [inline kge/job/eval.py:158]: Why is this done? Quit should be repsected, not changed.

File: kge/job/eval.py [hunk 7]
```diff
@@ -141,3 +146,66 @@ def create_from(
...
+
+        self._train_job = TrainingJob.create(
+            config=training_loss_eval_config,
+            parent_job=self,
+            dataset=dataset,
+            model=model,
+            forward_only=True,
+        )
+
+        # prepare training job
```
Comment (rgemulla) [inline kge/job/eval.py:171]: Move this to _prepare of this job

File: kge/job/eval.py [hunk 8]
```diff
@@ -141,3 +146,66 @@ def create_from(
...
+            split=self._train_job.config.get("train.split"),
+            epoch=self.epoch,
+        )
+
+        # run pre-epoch hooks (may modify trace)
+        for f in self.pre_epoch_hooks:
+            f(self)
+
+        # let's go
+        train_trace_entry = self._train_job.run_epoch()
```
Comment (rgemulla) [inline kge/job/eval.py:201]: metric computation is missing here. in particular, if the metric is not in the trace, the metric expr needs to be computed. the code that does this should be moved to our metric class

File: kge/job/train.py [hunk 9]
```diff
@@ -490,6 +523,7 @@ def run_epoch(self) -> Dict[str, Any]:
         # add results to trace entry
         self.current_trace["epoch"].update(
             dict(
+                sum_loss=sum_loss,
```
Comment (rgemulla) [inline kge/job/train.py:526]: I think we should drop sum_loss again. Since we already report the right avg loss (averaged over number of examples), the sum_loss can be comkputed easiliy outside (num_examples is in the trace, right?). The good thing about only having avg loss is that is makes more clear that this is our training objective (which matters to interpret regulraizitaion constants)

File: kge/job/eval.py (hunk 8 above)
Comment (rgemulla) [reply kge/job/eval.py:201]: My bad, this is invalid!

File: kge/job/eval.py (hunk 5 above)
Comment (rufex2001) [reply kge/job/eval.py:152]: Didn't even realize :)

File: kge/job/eval.py (hunk 6 above)
Comment (rufex2001) [reply kge/job/eval.py:158]: Got it from the old PR, thought it was discussed already. I guess if we want to simply evaluate a model, we don't need the output from the internal training job

File: kge/job/eval.py (hunk 7 above)
Comment (rufex2001) [reply kge/job/eval.py:171]: Right, where it should be!

File: kge/job/train.py (hunk 9 above)
Comment (rufex2001) [reply kge/job/train.py:526]: I Agree about clarity of training objective! Num_examples is not in the trace though, but will add it so sum loss can be computed from it

File: kge/job/eval.py (hunk 8 above)
Comment (rufex2001) [reply kge/job/eval.py:201]: But it does raise an interesting issue: what should be added to current_trace by this type of eval job? I guess simply the training job trace? Currently, it is "harcoded" to return avg loss and sum loss.

Comment (rufex2001) [reply kge/job/eval.py:201]: I guess we should add the event eval completed and done. Then custom metrics can be defined on the train trace

**Comment by rufex2001** [comment]
@rgemulla Ok, num_examples was there but called size. Renamed it to num_examples because I think that makes more sense than the notion of "epoch size". We do call it size at the batch level, which of course makes sense. If you want, I can revert this.

Also, not sure what keys to include in the trace produced by this eval job. I manually add avg_loss and num_examples from the internal training job for now. When adding everything, some things are overwritten, e.g. the event key, type key, epoch number key, etc. Also, this causes an error during the trace event, e.g. because there is already a key for job_id that comes from the already traced entry from the training job. We need to either decide which keys from the train job to include (avg cost too? avg penalty?), or which keys to exclude to prevent errors and overwritten keys. Ideally, it would be nice if all relevant keys from the train job were available to design custom metrics.

File: kge/job/eval.py (hunk 6 above)
Comment (rgemulla) [reply kge/job/eval.py]: I think we do, but if we don't then quiet should be set to True, right?

File: kge/job/eval.py (hunk 8 above)
Comment (rgemulla) [reply kge/job/eval.py:198]: good question. for now, we could trace the train entry separately (as we do for valid) and take over the loss, penalty, cost to the eval job.

**Comment by rufex2001** [comment]
@rgemulla Tracing done as suggested.

File: kge/job/train.py [hunk 10]
```diff
@@ -288,9 +312,12 @@ def run_epoch(self) -> Dict[str, Any]:
             epoch=self.epoch,
             split=self.train_split,
             batches=len(self.loader),
-            size=self.num_examples,
-            lr=[group["lr"] for group in self.optimizer.param_groups],
+            num_examples=self.num_examples,
```
Comment (rgemulla) [inline kge/job/train.py:315]: Since we use size everywhere else (e.g., for the batch size), I suggest to continue using "size" here.

**Review by rgemulla — COMMENTED** [review]
Other than this, looks good to me.

**Comment by rufex2001** [comment]
@rgemulla Fixed!

**Comment by rgemulla** [comment]
Alright, please add the CHANGELOG and merge
